### DISCUSSION TITLE
How to pass dynamic (the length of the list is not determined) arguments list from DAG parameters to a KubernetesPodOperator.

### DISCUSSION THREAD

**Discussion by mandarin801** [apache/airflow#36451] [Q&A] [answered]
Since I need to have a dynamic list of arguments in KubernetesPodOperator, I'm going to pass arguments list from DAG parameter to a KubernetesPodOperator.
The problem is that the arguments are passed as a single string, and a KubernetesPodOperator requires them to be passed as a list of strings. I've tried to use the following code, but as far as I can see, it doesn't work, because `"{{ params.args_list }}"` returns string with arguments in KubernetesPodOperator parameters only and does't work in other functions. I've also tried to use `json.loads`, and supportive function in PythonOpetator and xcom, but it didn't work either.

```python
from airflow import DAG
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator
from airflow.models.param import Param
import ast

default_args = {
    'owner': 'john_doe',
    'start_date': days_ago(1),
    'depends_on_past': False,
    'email': ['email@email.com'],
    'email_on_failure': True,
}

dag = DAG(
    'dag',
    default_args=default_args,
    schedule_interval=None,
    catchup=False,
    params={"args_list": Param(["arg1", "arg2"], type="array") },
)


k8s_task = KubernetesPodOperator(
    task_id='task',
    name="task",
    namespace='default',
    image="alpine:latest",
    cmds=["sh", "-c", "echo"],
    arguments=ast.literal_eval("{{ params.args_list }}"),  # Tried different things here
    in_cluster=True,
    get_logs=True,
    dag=dag,
    do_xcom_push=True
)
```

Did anyone have a similar problem? How did you solve it? Maybe there is a better way to pass a dynamic list of arguments to KubernetesPodOperator?

**Comment by mandarin801** [comment, marked as answer]
I've managed to make it working using Variables and PythonOperator. I've created a task that prepares arguments for KubernetesPodOperator and then I've passed them as op_args to KubernetesPodOperator.

```python
def prepare_arguments(params_string):
    return Variable.set("PARAMS", params_string)

prepare_args_task = PythonOperator(
    task_id='prepare_args',
    python_callable=prepare_arguments,
    op_args={"{{ params.args_list }}"},
    dag=dag,
)

params_list_string = Variable.get("PARAMS")
parmas_list = ast.literal_eval(params_list_string)

k8s_task = KubernetesPodOperator(
    task_id='task',
    name="task",
    namespace='default',
    image="alpine:latest",
    cmds=["sh", "-c", "echo"],
    arguments=parmas_list,
    in_cluster=True,
    get_logs=True,
    dag=dag,
    do_xcom_push=True
)
```

**Reply by potiuk** [reply]
This is wrong. Specifically if you re-run your task in the future, it will take last set value of the variable, not the one that was used originally - so you will run into all kinds of inconsistencies. 

Dynamic atask mapping is the only good solution that you should use https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/dynamic-task-mapping.html#dynamic-task-mapping

**Reply by mandarin801** [reply]
In my particular case it's totally ok, but thanks for your idea! It looks more safe and sophisticated.
